### PR TITLE
Lecture3 examples

### PR DESCRIPTION
**Pull request by ExtremeSkepticist**

### CODE REVIEW

File: 03/Dog.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+#include "Animal.h"
+
+class Dog : public Animal
+{
+public:
+	Dog();
+	Dog(int age, const char* size);
+	~Dog();
+
+	char* GetSize() const;
```
Comment (popekim) [inline 03/Dog.h]: please return const char*

File: 03/FunctionOverloadExample.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#include "Vector3.h"
+
+using namespace std;
+
+namespace samples
+{
+	void FunctionOverloadExample()
+	{
+		cout << "+------------------------------+" << endl;
+		cout << "|       Function Overloading Example       |" << endl;
```
Comment (popekim) [inline 03/FunctionOverloadExample.cpp]: does this line up properly?

File: 03/CopyConstructorExample.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+		cout << "|       Copy Constructor example       |" << endl;
+		cout << "+------------------------------+" << endl;
+
+		MyString myName("	John Doe");
+		MyString myNameCopy(myName);
+	
+		myName.PrintString();
+		myNameCopy.PrintString();
+
+		myName.~MyString();
```
Comment (popekim) [inline 03/CopyConstructorExample.cpp]: delete these two lines.

File: 03/MyString.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
+#include <iostream>
+#include <string>
+#include "MyString.h"
+
+using namespace std;
+
+MyString::MyString(const char* string)
+	: mCapacity(strlen(string))
```
Comment (popekim) [inline 03/MyString.cpp]: delete mcapacity

File: 03/MyString.cpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	, mSize(strlen(string))
```
Comment (popekim) [inline 03/MyString.cpp]: should be strlen + 1

File: 03/MyString.cpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+{
+	memcpy(mString, string, strlen(string));
```
Comment (popekim) [inline 03/MyString.cpp]: use mSize not strelen(string)

Comment (popekim) [inline 03/MyString.cpp]: have to do mString = new char[mSIze]; first

File: 03/MyString.cpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+MyString::MyString(const MyString& reference)
```
Comment (popekim) [inline 03/MyString.cpp]: reference -> string

File: 03/MyString.cpp [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	: mCapacity(strlen(reference.mString))
```
Comment (popekim) [inline 03/MyString.cpp]: : MyString(string.mString)
{}

File: 03/MyString.h [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+class MyString
+{
+public:
+	MyString(const char* string);
+	MyString(const MyString& reference);
```
Comment (popekim) [inline 03/MyString.h]: reference -> string

File: 03/MyString.h [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+	~MyString();
+
+	void PrintString();
+
+private:
+	const int STRING_SIZE = 9;
```
Comment (popekim) [inline 03/MyString.h]: delete

File: 03/MyString.h [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+	char* mString = new char[STRING_SIZE];
```
Comment (popekim) [inline 03/MyString.h]: char* mString;

File: 03/MyString.h [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+	int mCapacity;
```
Comment (popekim) [inline 03/MyString.h]: delete mCapacity

File: 03/Vector3.cpp [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+#include "Vector3.h"
+
+
+Vector3::Vector3(int x, int y)
+	: mX(x)
+	, mY(y)
+{
+}
+
+int Vector3::Multiply(const Vector3& reference) const
```
Comment (popekim) [inline 03/Vector3.cpp]: reference -> v or vector

File: 03/Vector3.cpp [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+{
+	return (reference.mX * reference.mY) + (mX * mY);
```
Comment (popekim) [inline 03/Vector3.cpp]: i want memberwise multiplication.. not dot product(and this is actually not even a dot product)

File: 03/OperatorOverloadExample.cpp [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+	void OperatorOverloadExample()
+	{
+		cout << "+------------------------------+" << endl;
+		cout << "|       Operator Overloading Example       |" << endl;
+		cout << "+------------------------------+" << endl;
+	
+		Vector4 vector1(3, 5);
+		Vector4 vector2(7, 9);
+
+		const int SCALRA = 8;
```
Comment (popekim) [inline 03/OperatorOverloadExample.cpp]: SCALAR

File: 03/Vector3.cpp (hunk 13 above)
Comment (popekim) [inline 03/Vector3.cpp]: your functions in vector classes look all wrong so i'm goign to writ edown what we want here


Function version
1. void Scale(const Vector3&);
2. void Scale(float s);
3. Vector3 Multiply(const Vector3&);  // do memberwise multiply
4. Vector3 Mutiply(float s);

operator version
1. void operator*=(const Vector3&);
2. void operator*=(float s);
3. Vector3 operator*(const Vector3&);  // do memberwise multiply
4. Vector3 operator*(float s);

File: 03/Vector4.cpp [hunk 16]
```diff
@@ -0,0 +1,12 @@
+#include "Vector4.h"
+
+Vector4::Vector4(int x, int y)
```
Comment (popekim) [inline 03/Vector4.cpp]: vector1, and vector2 please?

File: 03/InheritanceExample.cpp [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+		catPtr = &cat;
+		dogPtr = &dog;
+
+		cout << "Cat - " << "Age :" << catPtr->GetAge() << ", " << "Name : " << catPtr->GetName() << endl;
+		cout << "Dog - " << "Age : " << dogPtr->GetAge() << ", " << "Size : " << dogPtr->GetSize() << endl;
+
+		delete[] catPtr;
+		delete[] dogPtr;
+
+		cat.~Cat();
```
Comment (popekim) [inline 03/InheritanceExample.cpp]: delete these two lines

File: 03/InheritanceExample.cpp [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+		Cat* catPtr = new Cat;
+		Dog* dogPtr = new Dog;
+
+		catPtr = &cat;
+		dogPtr = &dog;
+
+		cout << "Cat - " << "Age :" << catPtr->GetAge() << ", " << "Name : " << catPtr->GetName() << endl;
+		cout << "Dog - " << "Age : " << dogPtr->GetAge() << ", " << "Size : " << dogPtr->GetSize() << endl;
+
+		delete[] catPtr;
```
Comment (popekim) [inline 03/InheritanceExample.cpp]: these delete array not object.. so no

File: 03/InheritanceExample.cpp [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+		cout << "|       Inheritance Example       |" << endl;
+		cout << "+------------------------------+" << endl;
+
+		Cat cat(3, "Kitty");
+		Dog dog(5, "Puppy");
+
+		Cat* catPtr = new Cat;
+		Dog* dogPtr = new Dog;
+
+		catPtr = &cat;
```
Comment (popekim) [inline 03/InheritanceExample.cpp]: memory leak already...

File: 03/InheritanceExample.cpp [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+#include "Dog.h"
+#include "InheritanceExample.h"
+
+using namespace std;
+
+namespace samples
+{
+	void InheritanceExample()
+	{
+		cout << "+------------------------------+" << endl;
```
Comment (popekim) [inline 03/InheritanceExample.cpp:13]: redo this example... there are memory leaks...

File: 03/Dog.cpp [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+
+using namespace std;
+
+Dog::Dog()
+{
+	cout << "Dog : constructor" << endl;
+}
+
+Dog::Dog(int age, const char* size)
+	:Animal(age)
```
Comment (popekim) [inline 03/Dog.cpp]: : Animal

File: 03/Cat.cpp [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+	strcpy(mName, name);
+}
+
+Cat::~Cat()
+{
+	delete[] mName;
+	cout << "Cat : destructor" << endl;
+}
+
+char* Cat::GetName() const
```
Comment (popekim) [inline 03/Cat.cpp]: const char*

File: 03/Cat.h [hunk 23]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+#include "Animal.h"
+
+class Cat : public Animal
+{
+public:
+	Cat();
```
Comment (popekim) [inline 03/Cat.h]: let's delete parameterless constructor?

File: 03/Animal.cpp [hunk 24]
```diff
@@ -0,0 +1,24 @@
+#include <iostream>
+#include "Animal.h"
+
+using namespace std;
+
+Animal::Animal()
+{
```
Comment (popekim) [inline 03/Animal.cpp]: : mAge(0)

File: 03/Dog.h [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+
+private:
+	char* mSize;
```
Comment (popekim) [inline 03/Dog.h]: dog has size . .cat has name... doesn' tmake huge sense...

so this example doesn't show proper OOP.... 

maybe we can find functions that's unique to cat .. and another functions taht's unique to dog?

**Review by popekim — CHANGES_REQUESTED** [review]

File: 03/Vector3.cpp (hunk 13 above)
Comment (ExtremeSkepticist) [reply 03/Vector3.cpp]: Just to be clear..

1. Scale functions does multiplying 2 power n to each member variable(mX, mY) and return the results
2. Multiply functions does multiplying the member variables to another object's member variables and return the results

Am I right?

Comment (popeskyboxlabs) [reply 03/Vector3.cpp]: 1. scale doesn't return anything it's void so it changes the member variables

mX *= s;
mY *= s;

would do

actually change float s to int s

2. doesn't change member variable at all

when you do 

3 * 5 = 15

3 and 5 doesn't change.

Comment (popekim) [reply 03/Vector3.cpp]: No

scale doesn't return anything it's void so it changes the member variables
mX *= s;
mY *= s;

would do

actually change float s to int s

doesn't change member variable at all
when you do

3 * 5 = 15

3 and 5 doesn't change.

**Pull request merged**